### PR TITLE
Support antlr as plan parser

### PR DESCRIPTION
**Pull request by longjiquan**
issue: #16658 

Co-authored-by: xaxys <tpnnghd@163.com>

Signed-off-by: dragondriver <jiquan.long@zilliz.com>

### CODE REVIEW

**Comment by czs007** [comment]
@DragonDriver  please move planparserv2 to internal/util

File: internal/util/typeutil/schema.go [hunk 1]
```diff
@@ -238,6 +238,11 @@ func IsFloatingType(dataType schemapb.DataType) bool {
 	}
 }
 
+// IsNumberType returns true if input is of number type, otherwise false.
+func IsNumberType(dataType schemapb.DataType) bool {
```
Comment (czs007) [inline internal/util/typeutil/schema.go:242]: isArithmetic

**Comment by DragonDriver** [comment]
rerun ut

File: internal/util/planparserv2/README.md [hunk 2]
```diff
@@ -0,0 +1,23 @@
+# Generate Parser with Antlr4
+
+## Install Antlr4
+
+Please follow [install antlr4](https://github.com/antlr/antlr4/blob/master/doc/go-target.md) to install the antlr tool.
+
+The version of antlr tool: `4.10.1`.
```
Comment (xiaofan-luan) [inline internal/util/planparserv2/README.md:7]: I can not run all unit test with 4.10.1 generated code.
4.9.3 works for me

Comment (xiaofan-luan) [reply internal/util/planparserv2/README.md:7]: curl -O https://www.antlr.org/download/antlr-4.9.3-complete.jar

File: internal/util/planparserv2/plan_parser_v2.go [hunk 3]
```diff
@@ -0,0 +1,105 @@
...
+		return nil, nil
+	}
+
+	ret := handleExpr(schema, exprStr)
+	predicate := getExpr(ret)
+	if predicate == nil {
+		return nil, fmt.Errorf("cannot parse expression: %s", exprStr)
+	}
+	if !typeutil.IsBoolType(predicate.dataType) {
+		return nil, fmt.Errorf("predicate is not a boolean expression")
```
Comment (xiaofan-luan) [inline internal/util/planparserv2/plan_parser_v2.go:48]: print the exprStr and predicate.dataType

File: internal/util/planparserv2/plan_parser_v2.go [hunk 4]
```diff
@@ -0,0 +1,105 @@
...
+
+func ParseExpr(schema *typeutil.SchemaHelper, exprStr string) (*planpb.Expr, error) {
+	if len(exprStr) <= 0 {
+		return nil, nil
+	}
+
+	ret := handleExpr(schema, exprStr)
+	predicate := getExpr(ret)
+	if predicate == nil {
+		return nil, fmt.Errorf("cannot parse expression: %s", exprStr)
```
Comment (xiaofan-luan) [inline internal/util/planparserv2/plan_parser_v2.go:45]: should print error in ret?

File: internal/util/planparserv2/plan_parser_v2_test.go [hunk 5]
```diff
@@ -0,0 +1,463 @@
+package planparserv2
```
Comment (xiaofan-luan) [inline internal/util/planparserv2/plan_parser_v2_test.go:1]: make sure all plan_parser_test.go expression test can be passed

File: internal/util/planparserv2/utils.go [hunk 6]
```diff
@@ -0,0 +1,315 @@
...
+		// 2 % a == 3
+		switch arithExpr.GetOp() {
+		case planpb.ArithOpType_Add, planpb.ArithOpType_Mul:
+			return combineBinaryArithExpr(op, arithExpr.GetOp(), rightExpr.GetInfo(), leftValue.GetValue(), valueExpr.GetValue()), nil
+		default:
+			return nil, fmt.Errorf("todo")
+		}
+	} else {
+		// (a + b) / 2 == 3
+		return nil, fmt.Errorf("complicated arithmetic operations are not supported")
```
[the remark this answers is not in the document]
Comment (xiaofan-luan) [reply internal/util/planparserv2/utils.go:239]: leave  it to later patch

File: internal/util/planparserv2/Plan.g4 [hunk 7]
```diff
@@ -0,0 +1,130 @@
+grammar Plan;
```
[the remark this answers is not in the document]
Comment (DragonDriver) [reply internal/util/planparserv2/Plan.g4:1]: OK, I'll fix refactor this.

File: internal/util/planparserv2/Plan.g4 [hunk 8]
```diff
@@ -0,0 +1,130 @@
...
+OR: '||' | 'or';
+
+BNOT: '~';
+NOT: '!' | 'not';
+
+IN: 'in';
+NIN: 'not in';
+EmptyTerm: '[]';
+
+BooleanConstant: 'true' | 'True' | 'false' | 'False';
```
[the remark this answers is not in the document]
Comment (DragonDriver) [reply internal/util/planparserv2/Plan.g4:69]: I made a test and original parser supports them. So they should be supported also.

File: internal/util/planparserv2/floating_comparision.go [hunk 9]
```diff
@@ -0,0 +1,9 @@
+package planparserv2
+
+import "math"
+
+const float64EqualityThreshold = 1e-9
+
+func almostEqual(a, b float64) bool {
```
[the remark this answers is not in the document]
Comment (DragonDriver) [reply internal/util/planparserv2/floating_comparision.go:7]: This also confuses me. At the beginning, I didn't write this function and I used `==` operator to do comparision, but I got a golint error.
In fact, I noticed that our original parser didn't support this situation. For example, `1.0 == 2.0` is not supported.

File: internal/util/planparserv2/README.md (hunk 2 above)
Comment (DragonDriver) [reply internal/util/planparserv2/README.md:7]: My fault, I'll fix the doc error.

File: internal/util/planparserv2/plan_parser_v2.go [hunk 10]
```diff
@@ -0,0 +1,105 @@
...
+	"github.com/antlr/antlr4/runtime/Go/antlr"
+)
+
+func handleExpr(schema *typeutil.SchemaHelper, exprStr string) interface{} {
+	if exprStr == "" {
+		return nil
+	}
+
+	inputStream := antlr.NewInputStream(exprStr)
+	lexer := parser.NewPlanLexer(inputStream)
```
[the remark this answers is not in the document]
Comment (DragonDriver) [reply internal/util/planparserv2/plan_parser_v2.go:21]: Yes, it's also the way that antlr recommend.

```golang
// NewPlanLexer produces a new lexer instance for the optional input antlr.CharStream.
//
// The *PlanLexer instance produced may be reused by calling the SetInputStream method.
// The initial lexer configuration is expensive to construct, and the object is not thread-safe;
// however, if used within a Golang sync.Pool, the construction cost amortizes well and the
// objects can be used in a thread-safe manner.
```

File: internal/util/planparserv2/plan_parser_v2.go [hunk 11]
```diff
@@ -0,0 +1,105 @@
...
+	stream := antlr.NewCommonTokenStream(lexer, antlr.TokenDefaultChannel)
+	parse := parser.NewPlanParser(stream)
```
[the remark this answers is not in the document]
Comment (DragonDriver) [reply internal/util/planparserv2/plan_parser_v2.go:23]: OK

File: internal/util/planparserv2/utils.go (hunk 6 above)
[the remark this answers is not in the document]
Comment (DragonDriver) [reply internal/util/planparserv2/utils.go:239]: Yeah, our execution backend didn't support this yet.

File: internal/util/planparserv2/floating_comparision.go (hunk 9 above)
[the remark this answers is not in the document]
Comment (DragonDriver) [reply internal/util/planparserv2/floating_comparision.go]: fixed.

File: go.mod [hunk 12]
```diff
@@ -8,6 +8,7 @@ require (
 	github.com/Jeffail/tunny v0.1.4 // indirect
 	github.com/StackExchange/wmi v1.2.1 // indirect
 	github.com/antonmedv/expr v1.8.9
+    github.com/antlr/antlr4/runtime/Go/antlr v0.0.0-20210826220005-b48c857c3a0e // indirect
```
Comment (cydrain) [inline go.mod:11]: need alignment

Comment (DragonDriver) [reply go.mod:11]: Already aligned in fact. Github render's issue caused this.

**Comment by xiaofan-luan** [comment]
Thanks for @DragonDriver and @xaxys for the impressive feature!

**Pull request merged**